src/adc.rs

```diff
@@ -0,0 +1,562 @@
+                    // stop and clear overrun events
+                    self.rb.cr.modify(|_, w| w.adstp().set_bit());
+                    self.rb.isr.modify(|_, w| w.ovr().clear_bit());
+
+                    self.rb.cfgr.modify(|_, w| w
+                        .cont().clear_bit()
+                        .ovrmod().clear_bit()
+                    );
+
+                    self.rb.sqr1.modify(|_, w| w.l().bits(0));
+
+                    self.operation_mode = Some(OperationMode::OneShot);
+                }
+
+                fn set_align(&self, align: Align) {
+                    self.rb.cfgr.modify(|_, w| w.align().bit(align.bitvalue()));
+                }
+
+                fn enable(&mut self) {
+                    self.rb.cr.modify(|_, w| w.aden().set_bit());
+                    while self.rb.isr.read().adrdy().bit_is_clear() {}
+                }
+
+                fn disable(&mut self) {
+                    self.rb.cr.modify(|_, w| w.aden().clear_bit());
+                }
+
+
+                /// Calibrate according to 15.3.8 in the Reference Manual
+                fn calibrate(&mut self) {
+                    if !self.rb.cr.read().advregen().is_enabled() {
+                        self.advregen_enable();
+                        self.wait_advregen_startup();
+                    }
+
+                    self.disable();
+
+                    self.rb.cr.modify(|_, w| w
+                        // NOTE: needs to be adopted if implementing differential input
+                        .adcaldif().clear_bit()
+                        .adcal()   .set_bit());
+
+                    while self.rb.cr.read().adcal().bit_is_set() {}
+                }
+
+
+                fn wait_adc_clk_cycles(&self, cycles: u32) {
+                    let adc_clk_cycle = self.clocks.hclk().0 / (self.ckmode as u32);
+                    asm::delay(adc_clk_cycle * cycles);
+                }
+
+                fn advregen_enable(&mut self){
+                    // need to go through intermediate first
+                    self.rb.cr.modify(|_, w| w.advregen().intermediate());
+                    self.rb.cr.modify(|_, w| w.advregen().enabled());
+
+                }
+
+                /// wait for the advregen to startup.
+                ///
+                /// This is based on the MAX_ADVREGEN_STARTUP_US of the device.
+                fn wait_advregen_startup(&self) {
+                        asm::delay((MAX_ADVREGEN_STARTUP_US * 1_000_000) / self.clocks.sysclk().0);
+                }
+
+                /// busy ADC read
+                fn convert_one(&mut self, chan: u8) -> u16 {
+                    self.ensure_oneshot();
+                    self.set_chan_smps(chan, SampleTime::default());
+                    self.select_single_chan(chan);
+
+                    self.rb.cr.modify(|_, w| w.adstart().set_bit());
+                    while self.rb.isr.read().eos().is_not_complete()
+                        {}
```

Comment:
Man, I really wish rustfmt would work inside macros. Feeling like a jackass nitpicking about all these format issues... But those braces should go on the previous line